### ISSUE TITLE
xref table missing object 0

### ISSUE THREAD

**Issue by briansorahan** [pdfcpu/pdfcpu#184]
# Version

```
pdfcpu ve
pdfcpu: v0.3.2 dev
```

# File

https://drive.google.com/open?id=1mq07MzAmgsybdE9Mw92rmVQcKtXebxL7

# What I Did

```
pdfcpu validate Bitwig\ Studio\ User\ Guide\ English.pdf
```

```
validating(mode=relaxed) Bitwig Studio User Guide English.pdf ...
Read: xRefTable failed: Free: object #0 not found.
```

# What I Expected

pdfcpu declares this to be a valid PDF.

The "Preview" application on my macbook is able to open this PDF.

**Comment by briansorahan** [comment]
Here is what pdfcpu logged when I ran it with -vv

```
Fatal: Free: object #0 not found.
github.com/pdfcpu/pdfcpu/pkg/pdfcpu.(*XRefTable).Free
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/pdfcpu/xreftable.go:250
github.com/pdfcpu/pdfcpu/pkg/pdfcpu.(*XRefTable).EnsureValidFreeList
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/pdfcpu/xreftable.go:491
github.com/pdfcpu/pdfcpu/pkg/pdfcpu.readXRefTable
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/pdfcpu/read.go:1216
github.com/pdfcpu/pdfcpu/pkg/pdfcpu.Read
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/pdfcpu/read.go:73
github.com/pdfcpu/pdfcpu/pkg/api.ReadContext
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:71
github.com/pdfcpu/pdfcpu/pkg/api.Validate
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:222
github.com/pdfcpu/pdfcpu/pkg/api.ValidateFile
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:268
github.com/pdfcpu/pdfcpu/pkg/cli.Validate
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/cli/cli.go:33
github.com/pdfcpu/pdfcpu/pkg/cli.Process
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/cli/process.go:104
main.process
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/main.go:105
main.handleValidateCommand
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/prepare.go:105
main.CommandMap.Handle
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/registry.go:85
main.main
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/main.go:61
runtime.main
	/usr/local/go/src/runtime/proc.go:203
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
Read: xRefTable failed
github.com/pdfcpu/pdfcpu/pkg/pdfcpu.Read
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/pdfcpu/read.go:74
github.com/pdfcpu/pdfcpu/pkg/api.ReadContext
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:71
github.com/pdfcpu/pdfcpu/pkg/api.Validate
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:222
github.com/pdfcpu/pdfcpu/pkg/api.ValidateFile
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/api/api.go:268
github.com/pdfcpu/pdfcpu/pkg/cli.Validate
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/cli/cli.go:33
github.com/pdfcpu/pdfcpu/pkg/cli.Process
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/pkg/cli/process.go:104
main.process
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/main.go:105
main.handleValidateCommand
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/prepare.go:105
main.CommandMap.Handle
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/registry.go:85
main.main
	/Volumes/BetterFS/github.com/pdfcpu/pdfcpu/cmd/pdfcpu/main.go:61
runtime.main
	/usr/local/go/src/runtime/proc.go:203
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
```

**Comment by hhrutter** [comment]
Thanks for reporting this.
I am going to see what's going on in this file.

**Comment by hhrutter** [comment]
This file's cross reference table does not contain object 0, which should be there representing the head of the linked list of empty objects.

I do have a workaround and will let you know when this is committed.
This may take some time though because I am busy otherwise at the moment.

Thanks for using pdfcpu 💚

**Comment by Cedric-Caballol-Shepy** [comment]
Hi,
I'm interested too for this fix. If you need more examples I have 2 "buggy" pdfs and I my scanner can produce some more ! 😉

**Comment by hhrutter** [comment]
Thanks I already have the fix.
Will be part of the next release!

BTW this seems the be the way how Apache FOP, quite an old tool based on XSLT,  creates PDFs in general. It assumes you don't need to create Object 0 when the xref's list of unused objects is empty - which is not valid per spec.

I cannot make this part of relaxed validation, since pdfcpu needs to validate the "freelist" much earlier when reading in the cross reference table way before validation.

**Comment by briansorahan** [comment]
Thx for looking at this and putting in a fix so quickly @hhrutter !

```
I cannot make this part of relaxed validation
```

means that `pdfcpu validate` in relaxed mode would declare this PDF valid?

**Comment by hhrutter** [comment]
This fix needs to be applied even before validation kicks in when pdfcpu reads the file in and builds up the cross reference table in memory.

**Comment by hhrutter** [comment]
Fixed with [v0.3.3](https://github.com/pdfcpu/pdfcpu/releases)
